power-sequencer/mihawk-cpld.cpp

```diff
@@ -0,0 +1,478 @@
+#include "config.h"
+
+#include "mihawk-cpld.hpp"
+
+#include "utility.hpp"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include <chrono>
+#include <elog-errors.hpp>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <org/open_power/Witherspoon/Fault/error.hpp>
+#include <phosphor-logging/elog.hpp>
+#include <phosphor-logging/log.hpp>
+#include <string>
+#include <xyz/openbmc_project/Common/Device/error.hpp>
+
+extern "C" {
+#include <i2c/smbus.h>
+#include <linux/i2c-dev.h>
+#include <linux/i2c.h>
+}
+
+// i2c bus & i2c slave address of Mihawk's CPLD
+#define busId 11
+#define slaveAddr 0x40
+
+// SMLink Status Register(PSU Register)
+const static constexpr size_t StatusReg = 0x05;
+
+// SMLink Status Register(Interrupt-control-bit Register)
+const static constexpr size_t StatusReg_1 = 0x20;
+
+// SMLink Status Register(Power-on error code Register)
+const static constexpr size_t StatusReg_2 = 0x21;
+
+using namespace std;
+namespace witherspoon
+{
+namespace power
+{
+const auto DEVICE_NAME = "MihawkCPLD"s;
+
+namespace fs = std::filesystem;
+using namespace phosphor::logging;
+
+using namespace sdbusplus::org::open_power::Witherspoon::Fault::Error;
+
+namespace device_error = sdbusplus::xyz::openbmc_project::Common::Device::Error;
+
+MihawkCPLD::MihawkCPLD(size_t instance, sdbusplus::bus::bus& bus) :
+    Device(DEVICE_NAME, instance), bus(bus)
+{
+}
+
+void MihawkCPLD::onFailure()
+{
+    bool psuError = checkPSUdevice();
+    bool poweronError = checkPoweronFault();
+
+    if (psuError)
+    {
+        // If PSU not present, report the error log & event.
+        report<PSUPresentError>();
+        std::cerr << "CPLD : PSU Present Error! \n";
+    }
+
+    // If the interrupt of power_on_error is switch on,
+    // read CPLD_register error code to analyze and report the error log
+    // & event.
+    if (poweronError)
+    {
+        std::cerr << "CPLD : Power_ON error! \n";
+        int errorcode;
+        errorcode = readFromCPLDPSUErrorCode(busId, slaveAddr);
+        if (errorcode == 0)
+        {
+            report<PowerOnErrorCPLDRegisterFail>();
+            std::cerr << "Power_on error : Read CPLD-register fail! \n";
+        }
+        else if (errorcode == 1)
```

Comment:
The value read by the readFromCPLDPSUErrorCode() function from the register of the CPLD means the cause of the error when **power on fail**. 
Because the readFromCPLDPSUErrorCode() function has only one field to read, there are only a  possible error.
The values read mean different reasons, for example,
If the value read is 01, it means PSU0 pgood fail.
If the value read is 02, it means PSU1 pgood fail.
If the value read is 0A (10), it means 240Va_Fault_H.

Reply:
ah, I get it.  Thanks.